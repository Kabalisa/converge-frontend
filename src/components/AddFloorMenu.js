import React from 'react';
import PropTypes from 'prop-types';
import { graphql } from 'react-apollo';
import { IconMenu, MenuItem, MenuDivider } from 'react-toolbox/lib/menu';
import AddFloorComponent from './AddFloor';
import { GET_ALL_OFFICES_QUERY } from '../graphql/queries/Offices';
import Spinner from './commons/Spinner';

/**
 * Renders the add office button
 *
 * @returns {JSX}
 */
const menuText = () => (
  <div className="addRoomBtn">
    <span>Add Floor</span>
  </div>
);

/**
 * Drop down list of offices to which floors can be added
 *
 * @returns {JSX}
 */
export class AddFloorMenu extends React.Component {
  static propTypes = {
    data: PropTypes.shape({
      allOffices: PropTypes.shape({
        office: PropTypes.array,
      }),
      loading: PropTypes.bool,
      error: PropTypes.object,
    }),
    refetch: PropTypes.func.isRequired,
  };

  static defaultProps = {
    data: {},
  };

  /**
 * Loops through all offices and returns each one of them as a menu item.
 *
 * @returns {JSX}
 */
  getOffice = () => {
    const { loading, allOffices: { offices } = {}, error } = this.props.data;
    if (loading) return <Spinner />;
    if (error) return <div>{error.message}</div>;
    const officeList = offices.filter(office => office.blocks.length !== 0);
    return (
      officeList.map(office => (
        <MenuItem key={office.id}>
          <AddFloorComponent
            theOffice={office.name}
            refetch={this.props.refetch}
            blocks={office.blocks}
          />
        </MenuItem>
      ))
    );
  }

  render() {
    return (
      <IconMenu className="add-room-menu" icon={menuText()}>
        <MenuItem caption="Select Office" disabled />
        <MenuDivider />
        {this.getOffice()}
      </IconMenu>
    );
  }
}

export default graphql(GET_ALL_OFFICES_QUERY, {
  options: () => ({
    variables: {
      page: 1,
      perPage: 50,
    },
  }),
})(AddFloorMenu);
