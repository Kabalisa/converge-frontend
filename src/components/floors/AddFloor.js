import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import toastr from 'toastr';
import MrmModal from '../commons/Modal';
import { Input, SelectInput as Select } from '../commons';
import ActionButtons from '../commons/ActionButtons';
import notification from '../../utils/notification';
import { ADD_FLOOR_MUTATION } from '../../graphql/mutations/Floors';

/**
 * Add Resource Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class AddFloor extends Component {
  static propTypes = {
    officeId: PropTypes.string,
    addFloor: PropTypes.func.isRequired,
    theOffice: PropTypes.string,
    currentPage: PropTypes.number,
    refetch: PropTypes.func.isRequired,
    blocks: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
    })).isRequired,
  };

  static defaultProps = {
    theOffice: '',
    currentPage: 1,
    officeId: '',
  };

  state = {
    floorName: '',
    blockId: '',
    closeModal: false,
    isCreating: false,
  };

  /**
   * Handles the closing of the modal
   *
   * @return {void}
  */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  /**
   * Ensures that the state is updated basing on the
   * user input
   *
   * @param {object} event
   * @param {number} capacity
   *
   * @returns {void}
   */
  handleInputChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  /**
   * It updates the state value of closeModal
   * to false whenever the modal closes
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   * 1. change isLoading state to it's opposite value
   * i.e true to false or vise verser
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isCreating: !this.state.isCreating,
    });
  }

  /**
   * 1. Submits the floor data to the backend
   * 2. Notifies the user about the response from the request
   * i.e whether it was a success or a failure
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleAddFloor = (event) => {
    event.preventDefault();
    const { blockId, floorName } = this.state;
    const { refetch, currentPage, officeId } = this.props;
    if (!floorName) {
      notification(toastr, 'error', 'Floor name is required')();
    } else if (!blockId && this.props.blocks) {
      notification(toastr, 'error', 'A block is required')();
    } else {
      this.toggleLoading();
      this.props
        .addFloor({
          variables: {
            name: floorName,
            blockId: blockId || officeId,
          },
        })
        .then(() => {
          this.toggleLoading();
          this.handleCloseModal();
          notification(
            toastr,
            'success',
            `${floorName} floor has been added successfully`,
          )();
          refetch({ page: currentPage });
        })
        .catch((err) => {
          this.toggleLoading();
          this.handleCloseModal();
          notification(toastr, 'error', err.graphQLErrors[0].message)();
        });
    }
  };

  /**
    * renders the modal for adding a floor.
    *
    * @returns {JSX}
    */
  render() {
    const {
      floorName, blockId, closeModal, isCreating,
    } = this.state;
    const { theOffice, blocks } = this.props;
    return (
      <MrmModal
        title="ADD FLOOR"
        buttonText={theOffice}
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
        className="add-office-modal"
        modalButton="add-button"
      >
        <form className="modal-form">
          <Input
            labelName="Floor Name"
            name="floorName"
            value={floorName}
            placeholder="Enter floor name"
            id="floorName"
            onChange={this.handleInputChange}
          />
          {blocks && <Select
            labelText="Select Block "
            name="blockId"
            id="block"
            value={blockId}
            onChange={this.handleInputChange}
            wrapperClassName="input-wrapper"
            placeholder="Select block"
            options={blocks}
          />
          }
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            actionButtonText="ADD FLOOR"
            isLoading={isCreating}
            onClickSubmit={this.handleAddFloor}
          />
        </form>
      </MrmModal>
    );
  }
}

AddFloor.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  blocks: PropTypes.any.isRequired,
};

export default compose(
  graphql(ADD_FLOOR_MUTATION, { name: 'addFloor' }),
)(AddFloor);
