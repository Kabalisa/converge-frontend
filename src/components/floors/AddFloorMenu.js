import React from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import { IconMenu, MenuItem, MenuDivider } from 'react-toolbox/lib/menu';
import AddFloorComponent from './AddFloor';
import Spinner from '../commons/Spinner';
import GET_ALL_BLOCKS from '../../graphql/queries/Blocks';

/**
 * Renders the add office button
 *
 * @returns {JSX}
 */
const menuText = () => (
  <div className="addRoomBtn">
    <span>Add Floor</span>
  </div>
);

/**
 * Drop down list of offices to which floors can be added
 *
 * @returns {JSX}
 */
export class AddFloorMenu extends React.Component {
  static propTypes = {
    allBlocks: PropTypes.shape({
      allBlocks: PropTypes.array,
      loading: PropTypes.bool,
      error: PropTypes.object,
    }),
    refetch: PropTypes.func.isRequired,
    currentPage: PropTypes.number,
  };

  static defaultProps = {
    allBlocks: {},
    currentPage: 1,
  };

  /**
 * Loops through all offices and returns each one of them as a menu item.
 *
 * @returns {JSX}
 */
  getBlocks = () => {
    const { loading, allBlocks, error } = this.props.allBlocks;
    if (loading) return <Spinner />;
    if (error) return <div>{error.message}</div>;
    return (
      allBlocks.map(block => (
        <MenuItem key={block.id}>
          <AddFloorComponent
            theOffice={block.name}
            refetch={this.props.refetch}
            blocks={block.offices.location.name === 'Nairobi' && allBlocks}
            currentPage={this.props.currentPage}
            officeId={block.id}
          />
        </MenuItem>
      ))
    );
  }

  render() {
    return (
      <IconMenu className="add-room-menu" icon={menuText()}>
        <MenuItem caption="Select Office" disabled />
        <MenuDivider />
        {this.getBlocks()}
      </IconMenu>
    );
  }
}

export default compose(
  graphql(GET_ALL_BLOCKS, {
    name: 'allBlocks',
  }),
)(AddFloorMenu);
