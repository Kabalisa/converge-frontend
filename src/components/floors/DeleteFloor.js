import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import toastr from 'toastr';
import ActionButtons from '../commons/ActionButtons';
import MrmModal from '../commons/Modal';
import { DELETE_FLOOR_MUTATION } from '../../graphql/mutations/Floors';
import notification from '../../utils/notification';
import '../../assets/styles/deleteModal.scss';

/**
 * Delete Floor Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class DeleteFloor extends Component {
  state = {
    closeModal: false,
    isDeleting: false,
  };

  /**
   * It closes a modal
   *
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  /**
   * Handles the state changes for the deleting floor modal
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   * Handles deleting floor
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleDeleteFloor = (event) => {
    event.preventDefault();
    this.toggleLoading();
    this.props
      .deleteFloor({
        variables: {
          floorId: this.props.floorId,
        },
      })
      .then(() => {
        this.toggleLoading();
        this.handleCloseModal();
        notification(
          toastr,
          'success',
          `'${this.props.floorName}' has been deleted successfully`,
        )();
      })
      .catch((err) => {
        this.toggleLoading();
        this.handleCloseModal();
        notification(toastr, 'error', err.graphQLErrors[0].message)();
      });
  };

  /**
   * 1. change isLoading state to it's opposite value
   * i.e true to false or vise verser
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isDeleting: !this.state.isDeleting,
    });
  }

  render() {
    const { closeModal, isDeleting } = this.state;

    return (
      <MrmModal
        className="delete-modal"
        title="DELETE FLOOR"
        buttonText="Delete"
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
      >
        <div className="delete-modal-content">
          <p id="confirm-msg">
            Are you sure you want to delete {`"${this.props.floorName}"`}?{' '}
            <br />
            This cannot be undone
          </p>
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            isLoading={isDeleting}
            actionButtonText="DELETE FLOOR"
            onClickSubmit={this.handleDeleteFloor}
          />
        </div>
      </MrmModal>
    );
  }
}

DeleteFloor.propTypes = {
  floorName: PropTypes.string.isRequired,
  floorId: PropTypes.string.isRequired,
  deleteFloor: PropTypes.func.isRequired,
};

export default compose(
  graphql(DELETE_FLOOR_MUTATION, {
    name: 'deleteFloor',
  }),
)(DeleteFloor);
