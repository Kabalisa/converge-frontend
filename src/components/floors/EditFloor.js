import React from 'react';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import { graphql } from 'react-apollo';
import MrmModal from '../commons/Modal';
import ActionButtons from '../commons/ActionButtons';
import { Input } from '../commons';
// import '../assets/styles/editoffice.scss';
import notification from '../../utils/notification';
import { EDIT_FLOOR_MUTATION } from '../../graphql/mutations/Floors';

/**
 * Edit Floor Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class EditFloor extends React.Component {
  state = {
    floorName: this.props.floorName,
    floorId: this.props.floorId,
    closeModal: false,
  };

  /**
   * Ensures that the state is updated basing
   * on the changes in the input fields
   *
   * @param {Object} target
   *
   * @returns {void}
   */
  handleInputChange = ({ target: { name, value } }) => {
    this.setState({ [name]: value });
  };

  /**
   * It closes a modal
   *
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  /**
   * It changes the state of the modal
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   * Handles editing floor
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleEditFloor = (event) => {
    event.preventDefault();
    const { floorId, floorName } = this.state;

    this.props
      .editFloor({
        variables: {
          floorId,
          name: floorName,
        },
      })
      .then(() => {
        notification(toastr, 'success', `${floorName} Floor has been updated successfully`)();
      })
      .catch((err) => {
        notification(toastr, 'error', err.graphQLErrors[0].message)();
      });
    this.handleCloseModal();
  };

  render() {
    const { floorName, closeModal } = this.state;

    return (
      <MrmModal
        title="EDIT FLOOR"
        buttonText="Edit"
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
        className="edit-office-modal"
        modalButtonClassName="edit-button"
      >
        <div className="modal-form">
          <Input
            labelName="Floor Name"
            labelClass="label1"
            inputClass="input1"
            name="floorName"
            value={floorName}
            id="floorName"
            onChange={this.handleInputChange}
            required
          />
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            onClickSubmit={this.handleEditFloor}
            actionButtonText="SAVE CHANGES"
          />
        </div>
      </MrmModal>
    );
  }
}

EditFloor.propTypes = {
  editFloor: PropTypes.func.isRequired,
  floorName: PropTypes.string.isRequired,
  floorId: PropTypes.string.isRequired,
};

export default graphql(EDIT_FLOOR_MUTATION, {
  name: 'editFloor',
})(EditFloor);
