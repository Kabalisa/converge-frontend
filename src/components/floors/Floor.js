import React from 'react';
import PropTypes from 'prop-types';
import DeleteFloor from './DeleteFloor'; //eslint-disable-line
import EditFloor from './EditFloor'; //eslint-disable-line

/**
 * Floor Component
 *
 * @param {Object} floor
 *
 * @returns {JSX}
 */
export const Floor = ({ floor, refetch }) => (
  <div className="table__row">
    <span>{floor.name}</span>
    <span>{floor.block}</span>
    <span>
      <EditFloor
        id="edit-modal"
        floorName={floor.name}
        floorId={floor.id}
        floorLocation={floor.office}
      />
      &nbsp;
      <DeleteFloor floorName={floor.name} id="delete-modal" floorId={floor.id} refetch={refetch} />
    </span>
  </div>
);

Floor.propTypes = {
  floor: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }).isRequired,
  refetch: PropTypes.func.isRequired,
};

export default Floor;
