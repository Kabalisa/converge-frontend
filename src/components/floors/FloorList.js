import React, { Component } from 'react';
import { graphql, compose } from 'react-apollo';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import { Floor } from './Floor';
import AddFloorMenuComp from './AddFloorMenu';
import Overlay from '../commons/Overlay';
import notification from '../../utils/notification';
import Spinner from '../commons/Spinner';
import { GET_PAGINATED_FLOORS_QUERY } from '../../graphql/queries/Floors';
import ColGroup from '../helpers/ColGroup';
import TableHead from '../helpers/TableHead';
import { formatFloorData } from '../../graphql/mappers/Floors';
import { GET_USER_ROLE } from '../../graphql/queries/People';
import { decodeTokenAndGetUserData } from '../../utils/Cookie';
import { saveItemInLocalStorage } from '../../utils/Utilities';
import MenuTitle from '../commons/MenuTitle';
import Pagination from '../commons/Pagination';
import DataNotFound from '../commons/DataNotFound';

/**
 * Floor List Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class FloorList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      allFloors: { ...props.data.allFloors },
      currentPage: 1,
      dataFetched: true, // true when there is an active internet connection.
      isFetching: false,
    };
  }

  componentWillReceiveProps(props) {
    const { allFloors } = props.data;
    this.setState({
      allFloors,
    });
  }

  /**
   * Handles the data displayed per page
   *
   * @param {string} perPage
   * @param {string } page
   *
   * @returns {Function}
   */
  handleData = (perPage, page) => {
    this.setState({ isFetching: true });
    /* istanbul ignore next */
    /* Reasoning: find explicit way of testing configuration options */
    this.props.data
      .fetchMore({
        variables: {
          page,
          perPage,
        },
        updateQuery: (prev, { fetchMoreResult }) => {
          this.setState({
            allFloors: fetchMoreResult.allFloors,
            currentPage: page,
          });
        },
      })
      .then(() => this.setState({ dataFetched: true, isFetching: false }))
      .catch(() => {
        this.setState({ dataFetched: false, isFetching: false });
        notification(
          toastr,
          'error',
          'You seem to be offline, check your internet connection.',
        )();
      });
  };

  render() {
    const {
      loading, error, refetch,
    } = this.props.data;
    const {
      isFetching, dataFetched, currentPage, allFloors,
    } = this.state;
    const { user } = this.props.user;

    if (loading) {
      return <Spinner />;
    }
    if (user) saveItemInLocalStorage('access', user.roles[0].id);
    if (error && error.message === 'GraphQL error: No more resources') return <DataNotFound />;
    if (error) return <div>{error.message}</div>;
    return (
      <div className="settings-resource">
        <div
          className={`settings-resource-control
        ${isFetching ? 'disabled-button' : null}`}
        >
          <MenuTitle title="Floors" />
          <AddFloorMenuComp refetch={refetch} />
        </div>
        <div className="settings-resource-list">
          {isFetching ? <Overlay id="floors-table" /> : null}
          <table className="test-One">
            <ColGroup />
            <TableHead titles={['Floor', 'Block', 'Action']} />
            <tbody>
              {
                allFloors && allFloors.floors.map(floor => (
                  <Floor
                    floor={formatFloorData(floor)}
                    key={floor.id}
                    refetch={refetch}
                  />
              ))}
            </tbody>
          </table>
        </div>
        <Pagination
          totalPages={allFloors.pages}
          hasNext={allFloors.hasNext}
          hasPrevious={allFloors.hasPrevious}
          handleData={this.handleData}
          currentPage={currentPage}
          isFetching={isFetching}
          dataFetched={dataFetched}
        />
      </div>
    );
  }
}

FloorList.propTypes = {
  data: PropTypes.shape({
    fetchMore: PropTypes.func.isRequired,
    refetch: PropTypes.func.isRequired,
    loading: PropTypes.bool,
    allFloors: PropTypes.object,
    id: PropTypes.number,
    name: PropTypes.string,
    blockId: PropTypes.number,
    block: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      offices: PropTypes.string,
    }),
    error: PropTypes.object,
  }).isRequired,
  user: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
};

const { UserInfo: userData } = decodeTokenAndGetUserData() || {};

export default compose(
  graphql(GET_PAGINATED_FLOORS_QUERY, {
    options: () => ({
      variables: {
        page: 1,
        perPage: 5,
      },
    }),
  }),
  graphql(GET_USER_ROLE, {
    name: 'user',
    options: /* istanbul ignore next */ () => ({
      variables: {
        email:
          process.env.NODE_ENV === 'test'
            ? 'sammy.muriuki@andela.com'
            : userData.email,
      },
    }),
  }),
)(FloorList);
