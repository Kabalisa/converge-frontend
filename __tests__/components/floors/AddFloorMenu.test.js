import React from 'react';
import { shallow } from 'enzyme';
import { AddFloorMenu } from '../../../src/components/floors/AddFloorMenu';

describe('Test the AddFloor component', () => {
  const props = {
    data: {
      allOffices: {
        offices: [
          {
            id: 1,
            name: 'Kampala',
            blocks: {
              name: 'blockA',
              id: 12,
            },
          },
        ],
      },
      loading: false,
      error: undefined,
    },
    refetch: jest.fn(),
  };
  const wrapper = shallow(<AddFloorMenu {...props} />);
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
