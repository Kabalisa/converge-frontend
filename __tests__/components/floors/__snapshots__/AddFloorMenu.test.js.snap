// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the AddFloor component should render correctly 1`] = `
<ThemedIconMenu
  className="add-room-menu"
  composeTheme="deeply"
  icon={
    <div
      className="addRoomBtn"
    >
      <span>
        Add Floor
      </span>
    </div>
  }
  iconRipple={true}
  mapThemrProps={[Function]}
  menuRipple={true}
  position="auto"
  selectable={false}
>
  <ThemedThemed
    caption="Select Office"
    composeTheme="deeply"
    disabled={true}
    mapThemrProps={[Function]}
    ripple={true}
    rippleCentered={false}
    rippleClassName=""
    rippleMultiple={true}
    rippleSpread={2}
  />
  <ThemedMenuDivider
    composeTheme="deeply"
    mapThemrProps={[Function]}
  />
  <ThemedThemed
    composeTheme="deeply"
    disabled={false}
    key="1"
    mapThemrProps={[Function]}
    ripple={true}
    rippleCentered={false}
    rippleClassName=""
    rippleMultiple={true}
    rippleSpread={2}
  >
    <Apollo(AddFloor)
      blocks={
        Array [
          Object {
            "id": 1,
            "name": "Kampala",
            "offices": Object {
              "id": "2",
              "location": Object {
                "name": "Nairobi",
              },
            },
          },
        ]
      }
      currentPage={1}
      officeId={1}
      refetch={[MockFunction]}
      theOffice="Kampala"
    />
  </ThemedThemed>
</ThemedIconMenu>
`;
